### DISCUSSION TITLE
Replace the generic Step function with Given/When/Then functions to match the majority of other languages?

### DISCUSSION THREAD

**Discussion by mirogta** [cucumber/godog#542] [Polls]
# Problem

This is how we register a step definition currently in `godog`:
```feature
…
  Scenario: Eat 5 out of 12          # features/godogs.feature:6
    Given there are 12 godogs
…
```

```go
…
ctx.Step(`^there are (\d+) godogs$`, thereAreGodogs)
…
```

Note that even though the step says `Given`, the actual implementation does not have `Given` there , but rather a generic `Step`.

# Proposal

I am sure there were good reasons to keep a generic `Step` function, but do these reasons still hold true today?

Why don't we mark it as deprecated and introduce 3 new functions for `Given`, `When` and `Then`?

**Disclaimer:** I'm also working on adding `go` and `godog` support to [cucumber/language-service](https://github.com/cucumber/language-service/issues/72) and it is a bit _tricky_ because of the generic `Step` function. Having three separate function would make it easier and more consistent with all the other languages.

## Evidence:

[Java](https://cucumber.io/docs/cucumber/step-definitions/?lang=java):
```
    @Given("I have {int} cukes in my belly")
    public void i_have_n_cukes_in_my_belly(int cukes) {
        System.out.format("Cukes: %n\n", cukes);
    }
```

[Javascript](https://cucumber.io/docs/cucumber/step-definitions/?lang=javascript)
```
Given('I have {int} cukes in my belly', function (cukes) {
  console.log(`Cukes: ${cukes}`)
});
```

[Ruby](https://cucumber.io/docs/cucumber/step-definitions/?lang=ruby)
```
Given('I have {int} cukes in my belly') do |cukes|
  puts "Cukes: #{cukes}"
end
```

[Kotlin](https://cucumber.io/docs/cucumber/step-definitions/?lang=kotlin)
```
        Given("I have {int} cukes in my belly") { cukes: Int ->
                println("Cukes: $cukes")
        }
```

[Scala](https://cucumber.io/docs/cucumber/step-definitions/?lang=scala)
```
    Given("I have {int} cukes in my belly") { cukes: Int =>
        println(s"Cukes: $cukes")
    }
```

[.NET](https://specflow.org/learn/bdd/)
```
[When(@"I search for '(.*)'"]
public void WhenIPerformASimpleSearchOn(string searchTerm) {
var controller = new CatalogController();
actionResult = controller.search(searchTerm);
}
```

[PowerShell](https://pester.dev/docs/v4/commands/GherkinStep)
```
Given 'we have a destination folder' {
    mkdir testdrive:\target -ErrorAction SilentlyContinue
    'testdrive:\target' | Should -Exist
}
```

[Rust](https://github.com/cucumber-rs/cucumber)
```
#[given(expr = "{word} is hungry")] // Cucumber Expression
async fn someone_is_hungry(w: &mut World, user: String) {
    sleep(Duration::from_secs(2)).await;
    
    w.user = Some(user);
}
```

**Comment by vearutop** [comment]
This makes sense to me. 

I think this would bring more clarity to steps definitions, as it will be harder to misunderstand/misuse steps that are intended for result assertion with steps that are invoking action or set initial state.

If that also helps discovery of steps with automated tools (e.g. language service), that's even better.

**Comment by mirogta** [comment]
Only 3 votes so far, but all "Yes". Raised as https://github.com/cucumber/godog/issues/545
